pkgs/development/compilers/zing/default.nix

```
@@ -0,0 +1,89 @@
+{ alsa-lib
+, autoPatchelfHook
+, fetchzip
+, lib
+, setJavaClassPath
+, stdenv
+, stdenvNoCC
+, xorg
+, zlib
+,
+}:
+stdenvNoCC.mkDerivation (finalAttrs: {
+  pname = "zing";
+  version = "24.01.0.0";
+
+  src = fetchzip {
+    x86_64-linux = {
+      url =
+        "https://cdn.azul.com/zing-zvm/ZVM24.01.0.0/zing24.01.0.0-4-jdk21.0.2-linux_x64.tar.gz";
+      hash =
+        "sha512-ymMHpC1pQaLXz4pEfgPqV6TSX/bbFuqrY8CmYZaXe8c28BpEwJYvti8wXyERN4tJbegWUC21V3gZ68aCd6FgbA==";
+    };
+    aarch64-linux = {
+      url =
+        "https://cdn.azul.com/zing-zvm/ZVM24.01.0.0/zing24.01.0.0-4-jdk21.0.2-linux_aarch64.tar.gz";
+      hash =
+        "sha512-JSUdDEew/z4DuqPJGFltu1KO+i9WENFibRL+ZJzZ5uso4Ib8wTkKZE0tYTgr6UcVhgH8UK1aOnO68gmBJwa+6A==";
+    };
+  }."${stdenvNoCC.hostPlatform.system}";
+
+  buildInputs = [ alsa-lib stdenv.cc.cc.lib xorg.libXrender xorg.libXtst zlib ];
+  nativeBuildInputs = [ autoPatchelfHook ];
+  propagatedBuildInputs = [ setJavaClassPath ];
+
+  installPhase = ''
+    runHook preInstall
+
+    cp -R ./ $out
+
+    ln -s $out/include/linux/*_md.h $out/include
+
+    mkdir -p $out/nix-support
+    cat <<EOF > $out/nix-support/setup-hook
+    if [ -z "\''${JAVA_HOME-}" ]; then export JAVA_HOME=$out; fi
+    EOF
+
+    runHook postInstall
+  '';
+
+  doInstallCheck = stdenvNoCC.buildPlatform.system
+    == stdenvNoCC.hostPlatform.system;
+  installCheckPhase = ''
+    runHook preInstallCheck
+
+    pushd "$(mktemp -d)"
+
+    echo ${
+      lib.escapeShellArg ''
+        public class HelloWorld {
+          public static void main(String[] args) {
+            System.out.println("Hello World");
+          }
+        }
+      ''
+    } > HelloWorld.java
+
+    $out/bin/javac HelloWorld.java
```

I think this looks a bit cleaner, but it's not very important.

```suggestion
  helloWorldJava = ''
    public class HelloWorld {
      public static void main(String[] args) {
        System.out.println("Hello World");
      }
    }
  '';
  passAsFile = [ "helloWorldJava" ];

  doInstallCheck = stdenvNoCC.buildPlatform.system
    == stdenvNoCC.hostPlatform.system;
  installCheckPhase = ''
    runHook preInstallCheck

    pushd "$(mktemp -d)"

    cp "$helloWorldJavaPath" HelloWorld.java
    $out/bin/javac HelloWorld.java
```